[spec] remove implementation details from enum spec

bishabosha opened the pull request:
- remove details that force all explicit extends clauses to have unique anonymous classes. Perhaps we can find a way in the future to desugar explicit extends clauses to the same implementation class
- also cleanup definitions of rules

sjrd commented on docs/_spec/05-classes-and-objects.md:
```diff
@@ -925,18 +925,15 @@ First, some terminology and notational conventions:
-     - one or more (possibly empty) parameter sections, denoted as `(´\mathit{ps}_1\,´)...(´\mathit{ps}_n´)`.
-  - _unparameterized_ enum classes have no type parameter sections and no parameter sections.
+  - _parameterized_ enum classes have at least one or more (possibly empty) term parameter clauses, denoted as `(´\mathit{ps}_1\,´)...(´\mathit{ps}_n´)`.
+  - _unparameterized_ enum classes have no term parameter clauses, but may optionally have a type parameter clause, denoted as `[´\mathit{tps}\,´]`.
```
That does not appear to be true:
```
scala> enum Foo { case Bar[T] }
-- [E004] Syntax Error: --------------------------------------------------------
1 |enum Foo { case Bar[T] }
  |                ^^^
  |                A case class must have at least one parameter list
  |
  | longer explanation available when compiling with `-explain`
1 error found
```

bishabosha replied:
this paragraph is about the base enum class, not a class case - hmm these terms are confusing

sjrd commented on docs/_spec/05-classes-and-objects.md:
```diff
@@ -970,13 +967,11 @@ enum Option[+T]:
-- a singleton case is mapped to a `val` definition, where
-  - Simple cases all share a single implementation class.
-  - Value cases will each be implemented by a unique class.
+- a class enum case is mapped to a `case class` member of enum class' companion object,
```
```suggestion
- a class enum case is mapped to a `case class` member of the enum class' companion object,
```

bishabosha replied:
done

sjrd commented on docs/_spec/05-classes-and-objects.md:
```diff
@@ -970,13 +967,11 @@ enum Option[+T]:
-  - Simple cases all share a single implementation class.
-  - Value cases will each be implemented by a unique class.
+- a class enum case is mapped to a `case class` member of enum class' companion object,
+- a singleton enum case is mapped to a `val` member of the enum class' companion object, implemented by a local class definition which may be shared between cases.
```
```suggestion
- a singleton enum case is mapped to a `val` member of the enum class' companion object, implemented by a local class definition. Whether that local class is shared with other singleton cases, and which ones, is left as an implementation detail.
```

bishabosha replied:
done

sjrd commented on docs/_spec/05-classes-and-objects.md:
````diff
@@ -989,106 +984,119 @@ transparent trait Enum extends Any, Product, Serializable:
 1.  An `enum` definition
     ```scala
-    enum ´E´ ... { <defs> <cases> }
+    enum ´E´ <type-params> <value-params> { <defs> <cases> }
````
```suggestion
    enum ´E´ <type-params> <value-params> extends <parents> { <defs> <cases> }
```

bishabosha replied:
done

sjrd commented on docs/_spec/05-classes-and-objects.md:
````diff
@@ -989,106 +984,119 @@ transparent trait Enum extends Any, Product, Serializable:
    ```
    Any modifiers or annotations on the original case extend to all expanded cases.
-   This result is then further rewritten by either (3 or 4).
+   <p>This result is then further rewritten by either (3 or 4).</p>
````
Does that actually render in a meaningful way?

bishabosha replied:
I thought it looks nicer with the clean break, there's no way to do that in pure markdown without breaking the numbered list

bishabosha commented on docs/_spec/05-classes-and-objects.md:
````diff
@@ -989,106 +984,119 @@ transparent trait Enum extends Any, Product, Serializable:
+   case ´C´(´\mathit{ps}_1\,´)...(´\mathit{ps}_n´)
    ```
-   expands to
+   of an unparameterized enum `´E´[´\mathit{tps}´]` with type parameters expands to
````
we can work on the exact terminology, but these desugaring rules where the extends clause is inferred requires that the base enum has no term parameters - so I was calling this an "unparameterized enum". Maybe you could suggest a better term than "enum without term parameters, but has type parameters"

sjrd replied:
Ah, OK, that's the same misunderstanding as above. It would be good to find a better name because "parameterized" refers to *type* parameters everywhere else in the spec. But it's not an issue with this PR's changes, so we can defer that.

bishabosha replied:
ok then this is to revisit

bishabosha commented:
@sjrd addressed the comments, we should probably review individual definition words, such as "enum class" and "unparameterized" later, as these seem to cause confusion unless you pay attention to every detail
